tokio-quiche/examples/async_http3_server/server.rs

```diff
@@ -0,0 +1,268 @@
+            H3Event::ConnectionError(err) => QuicResult::Err(Box::new(err)),
+
+            // The connection has shutdown.
+            H3Event::ConnectionShutdown(err) => {
+                let error = err.map_or_else(
+                    || Box::new(quiche::h3::Error::Done) as BoxError,
+                    |e| Box::new(e),
+                );
+
+                QuicResult::Err(error)
+            },
+
+            // Received headers for a new stream from the H3Driver.
+            H3Event::IncomingHeaders(headers) => {
+                self.handle_incoming_headers(headers).await;
+
+                Ok(())
+            },
+
+            _ => {
+                log::info!("received unknown event: {event:?}");
+                Ok(())
+            },
+        }
+    }
+
+    /// Respond to the request corresponding to the [`IncomingH3Headers`].
+    ///
+    /// This function transforms the incoming headers into a [`Request`],
+    /// creating the proper response body if requested. It then spawns a
+    /// Tokio task which calls the `service_fn` on the [`Request`].
+    async fn handle_incoming_headers(&mut self, headers: IncomingH3Headers) {
+        let IncomingH3Headers {
+            headers: list,
+            send: mut frame_sender,
+            ..
+        } = headers;
+
+        let Ok((uri_builder, req_builder)) = convert_headers(list) else {
+            Self::end_stream(&mut frame_sender).await;
+            return;
+        };
+
+        let uri = uri_builder.build().expect("can't build uri");
+        let req = req_builder.uri(uri).body(()).expect("can't build request");
+
+        let service_fn = Arc::clone(&self.service_fn);
+
+        tokio::spawn(async move {
+            Self::handle_request(service_fn, req, frame_sender).await;
+        });
+    }
+
+    /// Get a [`Response`] for a [`Request`] by calling the `service_fn`.
+    ///
+    /// The `frame_sender` parameter connects back to tokio-quiche `H3Driver`,
+    /// which communicates the data back to quiche.
+    async fn handle_request(
+        service_fn: Arc<S>, req: Request<()>,
+        mut frame_sender: OutboundFrameSender,
+    ) {
+        let res = service_fn(req).await;
+
+        // Convert the result of the `service_fn` into headers and a body which
+        // can be transmitted to tokio-quiche.
+        let (h3_headers, body) = convert_response(res);
+        let _ = frame_sender
+            .send(OutboundFrame::Headers(h3_headers, None))
+            .await;
+
+        send_body(body, frame_sender).await;
+    }
+
+    /// End the stream by sending STOP_SENDING and RESET_STREAM frames to the
+    /// client.
+    async fn end_stream(frame_sender: &mut OutboundFrameSender) {
+        let _ = frame_sender.send(OutboundFrame::PeerStreamError).await;
+    }
+}
+
+/// Convert a list of [Header]s into a request object which can be processed by
+/// the [`Server`].
+///
+/// This isn't complete or robust by any means.
```

Comment:
I find this a bit confusing. Could you rephrase this to say what is missing or what is supported.